### PR TITLE
ADM-709:[backend][docs]:Verify buildkite and obtain buildkite data with new API

### PR DESCRIPTION
**Pull request by Andrea2000728**
## Summary

add two api:

1.Verify the BuildKite token
2.getBuildKiteInfo

## Before

_Description_

**Screenshots**
If applicable, add screenshots to help explain behavior of your code.

## After

_Description_

**Screenshots**
If applicable, add screenshots to help explain behavior of your code.

## Note

_Null_

### CODE REVIEW

File: backend/src/main/java/heartbeat/controller/pipeline/PipelineController.java [hunk 1]
```diff
@@ -35,6 +36,26 @@ public BuildKiteResponseDTO getBuildKiteInfo(@PathVariable String pipelineType,
...
 
+	@PostMapping("/{pipelineType}/verify")
+	public ResponseEntity<Void> verifyBuildKiteToken(@PathVariable String pipelineType,
+			@Valid @RequestBody TokenParam tokenParam) {
+		buildKiteService.getBuildKiteVerify(tokenParam.getToken());
+		return ResponseEntity.noContent().build();
+	}
+
+	@PostMapping("/{pipelineType}/info")
+	public ResponseEntity<BuildKiteResponseDTO> fetchBuildKiteInfo(@PathVariable String pipelineType,
```
Comment (guzhongren) [inline backend/src/main/java/heartbeat/controller/pipeline/PipelineController.java:47]: pipelineType to a enum

Comment (guzhongren) [reply backend/src/main/java/heartbeat/controller/pipeline/PipelineController.java:47]: and remove `/info` in the path

File: backend/src/main/java/heartbeat/service/pipeline/buildkite/BuildKiteService.java [hunk 2]
```diff
@@ -300,4 +299,59 @@ private List<DeployInfo> getBuildsByState(List<BuildKiteBuildInfo> buildInfos,
...
+
+	public BuildKiteResponseDTO getBuildKiteInfo(PipelineParam pipelineParam) {
+		try {
+			String buildKiteToken = "Bearer " + pipelineParam.getToken();
+			log.info("Start to query BuildKite organizations by token");
+			List<BuildKiteOrganizationsInfo> buildKiteOrganizationsInfo = buildKiteFeignClient
+				.getBuildKiteOrganizationsInfo(buildKiteToken);
+			log.info("Successfully query BuildKite organizations by token, slug: {}", buildKiteOrganizationsInfo);
+
+			log.info("Start to query buildKite pipelineInfo by organizations slug: {}", buildKiteOrganizationsInfo);
```
Comment (guzhongren) [inline backend/src/main/java/heartbeat/service/pipeline/buildkite/BuildKiteService.java:330]: buildKite -> BuildKite

File: backend/src/main/java/heartbeat/service/pipeline/buildkite/BuildKiteService.java [hunk 3]
```diff
@@ -300,4 +299,59 @@ private List<DeployInfo> getBuildsByState(List<BuildKiteBuildInfo> buildInfos,
...
+			List<Pipeline> buildKiteInfoList = buildKiteOrganizationsInfo.stream()
+				.flatMap(org -> buildKiteFeignClient
+					.getPipelineInfo(buildKiteToken, org.getSlug(), "1", "100", pipelineParam.getStartTime(),
+							pipelineParam.getEndTime())
+					.stream()
+					.map(pipeline -> PipelineTransformer.fromBuildKitePipelineDto(pipeline, org.getSlug(),
+							org.getName())))
+				.collect(Collectors.toList());
+			log.info("Successfully get buildKite pipelineInfo, slug:{}, pipelineInfoList size:{}",
```
Comment (guzhongren) [inline backend/src/main/java/heartbeat/service/pipeline/buildkite/BuildKiteService.java:339]: buildKite -> BuildKite

File: backend/src/test/java/heartbeat/controller/pipeline/BuildKiteControllerTest.java [hunk 4]
```diff
@@ -116,4 +119,70 @@ void shouldReturnNoContentIfNoStepsWhenCallBuildKite() throws Exception {
 		assertThat(response.getContentAsString()).isEqualTo("");
 	}
 
+	@Test
+	void shouldReturnNoContentIfTokenCorrectWhenCallBuildKiteMockServer() throws Exception {
```
Comment (guzhongren) [inline backend/src/test/java/heartbeat/controller/pipeline/BuildKiteControllerTest.java:123]: why it is 204 when calling BuildKiteMockServer?

File: backend/src/test/java/heartbeat/controller/pipeline/BuildKiteControllerTest.java [hunk 5]
```diff
@@ -116,4 +119,70 @@ void shouldReturnNoContentIfNoStepsWhenCallBuildKite() throws Exception {
...
+				.content(mapper.writeValueAsString(tokenParam)))
+			.andExpect(status().isNoContent())
+			.andReturn()
+			.getResponse();
+
+		assertThat(response.getContentAsString()).isEqualTo("");
+	}
+
+	@Test
+	void shouldReturnPipelineInfoIfTokenCorrectWhenCallBuildKiteMockServer() throws Exception {
```
Comment (guzhongren) [inline backend/src/test/java/heartbeat/controller/pipeline/BuildKiteControllerTest.java:139]: BuildKiteMockServer??
no `if` in the test description. should given when

File: backend/src/test/java/heartbeat/controller/pipeline/BuildKiteControllerTest.java [hunk 6]
```diff
@@ -116,4 +119,70 @@ void shouldReturnNoContentIfNoStepsWhenCallBuildKite() throws Exception {
...
+		ObjectMapper mapper = new ObjectMapper();
+		List<Pipeline> pipelines = mapper.readValue(
+				new File("src/test/java/heartbeat/controller/pipeline/pipelineInfoData.json"), new TypeReference<>() {
+				});
+		BuildKiteResponseDTO buildKiteResponseDTO = BuildKiteResponseDTO.builder().pipelineList(pipelines).build();
+		PipelineParam pipelineParam = PipelineParam.builder()
+			.token("test_token")
```
Comment (guzhongren) [inline backend/src/test/java/heartbeat/controller/pipeline/BuildKiteControllerTest.java:146]: use fixture, not custom a variable everywhere

Comment (guzhongren) [reply backend/src/test/java/heartbeat/controller/pipeline/BuildKiteControllerTest.java:146]: and other positions

File: backend/src/test/java/heartbeat/controller/pipeline/BuildKiteControllerTest.java [hunk 7]
```diff
@@ -116,4 +119,70 @@ void shouldReturnNoContentIfNoStepsWhenCallBuildKite() throws Exception {
...
+			.build();
+
+		when(buildKiteService.getBuildKiteInfo(any())).thenReturn(buildKiteResponseDTO);
+
+		MockHttpServletResponse response = mockMvc
+			.perform(post("/pipelines/buildKite/info").contentType(MediaType.APPLICATION_JSON)
+				.content(mapper.writeValueAsString(pipelineParam)))
+			.andExpect(status().isOk())
+			.andReturn()
+			.getResponse();
```
Comment (guzhongren) [inline backend/src/test/java/heartbeat/controller/pipeline/BuildKiteControllerTest.java:158]: given a `new line` between `given when then`
refactor other test cases

File: backend/src/test/java/heartbeat/controller/pipeline/BuildKiteControllerTest.java [hunk 8]
```diff
@@ -116,4 +119,70 @@ void shouldReturnNoContentIfNoStepsWhenCallBuildKite() throws Exception {
...
+	}
+
+	@Test
+	void shouldReturnNoContentIfPipelineInfoIsNullWhenCallBuildKiteMockServer() throws Exception {
+		ObjectMapper mapper = new ObjectMapper();
+		BuildKiteResponseDTO buildKiteResponseDTO = BuildKiteResponseDTO.builder()
+			.pipelineList(Collections.emptyList())
+			.build();
+		PipelineParam pipelineParam = PipelineParam.builder()
+			.token("test_token")
```
Comment (guzhongren) [inline backend/src/test/java/heartbeat/controller/pipeline/BuildKiteControllerTest.java:172]: refactor...

File: backend/src/test/java/heartbeat/service/pipeline/buildkite/BuildKiteServiceTest.java [hunk 9]
```diff
@@ -551,4 +552,101 @@ void shouldReturnStepBeforeEndStepsGivenStepsArray() {
...
+			.scopes(List.of("read_builds", "read_organizations", "read_pipelines"))
+			.build();
+
+		when(buildKiteFeignClient.getTokenInfo(any())).thenReturn(buildKiteTokenInfo);
+		buildKiteService.getBuildKiteVerify("mock_token");
+
+		verify(buildKiteFeignClient, times(1)).getTokenInfo(anyString());
+	}
+
+	@Test
```
Comment (guzhongren) [inline backend/src/test/java/heartbeat/service/pipeline/buildkite/BuildKiteServiceTest.java:567]: given a new line between given when then

File: backend/src/test/java/heartbeat/service/pipeline/buildkite/BuildKiteServiceTest.java [hunk 10]
```diff
@@ -551,4 +552,101 @@ void shouldReturnStepBeforeEndStepsGivenStepsArray() {
...
+		ObjectMapper mapper = new ObjectMapper();
+		List<BuildKitePipelineDTO> pipelineDTOS = mapper.readValue(
+				new File("src/test/java/heartbeat/controller/pipeline/buildKitePipelineInfoData.json"),
+				new TypeReference<>() {
+				});
+
+		when(buildKiteFeignClient.getBuildKiteOrganizationsInfo(any()))
+			.thenReturn(List.of(BuildKiteOrganizationsInfo.builder().name("XXXX").slug("XXXX").build()));
+		when(buildKiteFeignClient.getPipelineInfo("Bearer test_token", "XXXX", "1", "100", "startTime", "endTime"))
+			.thenReturn(pipelineDTOS);
```
Comment (guzhongren) [inline backend/src/test/java/heartbeat/service/pipeline/buildkite/BuildKiteServiceTest.java:604]: those part should be in `when`, combine it

File: backend/src/test/java/heartbeat/service/pipeline/buildkite/BuildKiteServiceTest.java [hunk 11]
```diff
@@ -12,6 +12,7 @@
 import static org.mockito.ArgumentMatchers.any;
 import static org.mockito.ArgumentMatchers.anyString;
 import static org.mockito.Mockito.mock;
+import static org.mockito.Mockito.times;
```
Comment (guzhongren) [inline backend/src/test/java/heartbeat/service/pipeline/buildkite/BuildKiteServiceTest.java:15]: refactor this file with `given when then`

File: docs/src/content/docs/en/spikes/tech-spikes-split-verify-of-buildkite.mdx [hunk 12]
```diff
@@ -189,6 +190,7 @@ responses:{
       ]
     }]
 }
+当 pipelineList 为空时， responses 204
```
Comment (guzhongren) [inline docs/src/content/docs/en/spikes/tech-spikes-split-verify-of-buildkite.mdx:193]: covert it into english

**Review by guzhongren — CHANGES_REQUESTED** [review]
refactor with the comments

File: backend/src/test/java/heartbeat/controller/pipeline/BuildKiteControllerTest.java (hunk 4 above)
Comment (Andrea2000728) [reply backend/src/test/java/heartbeat/controller/pipeline/BuildKiteControllerTest.java:123]: When verify buildKite token successfully, we agreed to use 204 as the response. Also I change the BuildKiteMockServer to BuildKite

File: backend/src/main/java/heartbeat/controller/pipeline/PipelineController.java (hunk 1 above)
Comment (Andrea2000728) [reply backend/src/main/java/heartbeat/controller/pipeline/PipelineController.java:47]: "/{xxxxType}/info" We have designed this API to all follow this form （pipeline board source）. also the old api hasn't been removed yet, we can distinguish by this.

**Comment by Andrea2000728** [comment]
> refactor with the comments

all done

**Review by BoBoDai — APPROVED** [review]

**Pull request merged**